File: C/regular_expression_match.c
// https://leetcode.com/problems/regular-expression-matching/
/*
Given an input string (s) and a pattern (p), implement regular expression
 matching with support for '.' and '*' where: 

'.' Matches any single character.​​​​
'*' Matches zero or more of the preceding element.
The matching should cover the entire input string (not partial).

Example 1:
Input: s = "aa", p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".

Example 2:
Input: s = "aa", p = "a*"
Output: true
Explanation: '*' means zero or more of the preceding element, 'a'. Therefore, by repeating 'a' once, it becomes "aa".

Example 3:
Input: s = "ab", p = ".*"
Output: true
Explanation: ".*" means "zero or more (*) of any character (.)".

Example 4:
Input: s = "aab", p = "c*a*b"
Output: true
Explanation: c can be repeated 0 times, a can be repeated 1 time. Therefore, it matches "aab".

Example 5:
Input: s = "mississippi", p = "mis*is*p*."
Output: false
 
Constraints:
0 <= s.length <= 20
0 <= p.length <= 30
s contains only lowercase English letters.
p contains only lowercase English letters, '.', and '*'.
It is guaranteed for each appearance of the character '*', there will be a previous valid character to match.
*/

// Note - Do check the related wildcard_pattern_match.c
// Below isMatch1(...) will fail for test case : S = "spkm", P = "s*p*k*s*m"
// Best solution is DP sol. at - regular_expression_match.py
// Below isMatch(...) is recursive but correct.

#include <string.h>
#define false 0
#define true 1

typedef int bool;

bool isMatch1(char * s, char * p){
    int l1 = strlen(s);
    int l2 = strlen(p);
    int dot_astric_index_p = -1;
    int chr_astric_index_p = -1;
    int chr1_index_to_ignore = -1;
    int chr2_index_to_ignore = -1;
    int i = 0;
    int j = 0;
    while(i<l1){
        char a = s[i];
        char b = (j<l2)?p[j]:'\0';
        if(j+1<l2 && p[j+1]=='*')
            b = '-';
        if(b=='-'){
            if(p[j]=='.'){
                dot_astric_index_p = j;
                chr1_index_to_ignore = i;
                chr_astric_index_p = chr2_index_to_ignore = -1;
            }else if(a==p[j]){
                chr_astric_index_p = j;
                chr2_index_to_ignore = i;
            }
            j+=2;
        }else if(a==b || b=='.'){
            i+=1;
            j+=1;
        }else{
            if(chr_astric_index_p!=-1){
                i = ++chr2_index_to_ignore;
                j = chr_astric_index_p+2;
                if(s[chr2_index_to_ignore]!=p[chr_astric_index_p]){
                    chr2_index_to_ignore=chr_astric_index_p=-1;
                }
            }else if(dot_astric_index_p!=-1){
                i = ++chr1_index_to_ignore;
                j = dot_astric_index_p+2;
            }else{
                return false;
            }
        }
    }
    while(j<l2){
        if(j+1<l2 && p[j+1]!='*') return false;
        else if(j+1>=l2) return false;
        j+=2;
    }
    return true;
}

// Correct sol. below.
bool check(char *pat, int pat_len, int j){
    while(j<pat_len){
        if(j+1>=pat_len || pat[j+1]!='*') return false;
        j+=2;
    }
    return true;
}

bool match_txt_with_pattern(char *s, int s_len, char *p, int p_len, int i, int j){
    // printf("s_len=%d p_len=%d i=%d j=%d\n",s_len, p_len, i, j);
    // if(i==s_len) printf("check=%d\n",check(s, s_len, j));
    if(i==s_len) return check(p, p_len, j);
    if(j==p_len) return false;
    char txt = s[i];
    char pat = p[j];
    char next_pat = (j+1>=p_len)?'\0':p[j+1];
    // printf("txt=%c pat=%c next_pat=%c\n", txt, pat, next_pat);
    if(next_pat!='*'){
        if(pat=='.' || txt==pat) return match_txt_with_pattern(s, s_len, p, p_len, i+1, j+1);
        return false;
    }else{
        bool skip_pat = match_txt_with_pattern(s, s_len, p, p_len, i, j+2);
        if(skip_pat) return true;
        if(pat=='.' || txt==pat) return match_txt_with_pattern(s, s_len, p, p_len, i+1, j);
        return false;
    }
}

bool isMatch(char * s, char * p){
    int s_len = strlen(s);
    int p_len = strlen(p);
    return match_txt_with_pattern(s, s_len, p, p_len, 0, 0);
}

File: Python/regular_expression_match.py
# https://leetcode.com/problems/regular-expression-matching/
"""
Given an input string (s) and a pattern (p), implement regular expression
 matching with support for '.' and '*' where: 

'.' Matches any single character.​​​​
'*' Matches zero or more of the preceding element.
The matching should cover the entire input string (not partial).

Example 1:
Input: s = "aa", p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".

Example 2:
Input: s = "aa", p = "a*"
Output: true
Explanation: '*' means zero or more of the preceding element, 'a'. Therefore, by repeating 'a' once, it becomes "aa".

Example 3:
Input: s = "ab", p = ".*"
Output: true
Explanation: ".*" means "zero or more (*) of any character (.)".

Example 4:
Input: s = "aab", p = "c*a*b"
Output: true
Explanation: c can be repeated 0 times, a can be repeated 1 time. Therefore, it matches "aab".

Example 5:
Input: s = "mississippi", p = "mis*is*p*."
Output: false
 
Constraints:
0 <= s.length <= 20
0 <= p.length <= 30
s contains only lowercase English letters.
p contains only lowercase English letters, '.', and '*'.
It is guaranteed for each appearance of the character '*', there will be a previous valid character to match.
"""
class Pat:
    def __init__(self, char, is_astric, is_nullable):
        self.char = char
        self.is_astric = is_astric
        self.is_nullable = is_nullable
    
    def __str__(self):
        return self.char + ' ' + str(self.is_astric) + ' ' + str(self.is_nullable)
    
    def __repr__(self):
        return self.char + ':' + str(self.is_astric) + ':' + str(self.is_nullable)

class Solution(object):
    def compress_pattern(self, pattern):
        pat_len = len(pattern)
        op = list()
        is_nullable = True
        i = pat_len-1
        while i>=0:
            if pattern[i]=='*' and i-1>=0:
                op.insert(0, Pat(pattern[i-1], True, is_nullable))
                i-=2
            else:
                is_nullable = False
                op.insert(0, Pat(pattern[i], False, is_nullable))
                i-=1
        # print(op)
        return op

    def is_matching(self, s, t_p):
        s_len = len(s)
        p_len = len(t_p)
        if s_len==0:
            return True if p_len==0 else t_p[0].is_nullable
        if p_len==0:
            return False
        op = [False for k in range(s_len)]
        for i in range(p_len-1, -1, -1):
            prev = False
            for j in range(s_len-1, -1, -1):
                t_prev = op[j]
                pat = t_p[i].char
                txt = s[j]
                is_match = (pat=='.' or txt==pat)
                if t_p[i].is_astric == False:
                    if not is_match:
                        op[j] = False
                    else:
                        if i==p_len-1 and j==s_len-1: op[j] = True
                        elif j==s_len-1: op[j] = t_p[i+1].is_nullable
                        else: op[j] = prev
                else:
                    skip_pat = False if (i==p_len-1) else op[j]
                    use_txt  = is_match and (op[j+1] if (j<s_len-1) else t_p[i].is_nullable)
                    op[j] = (skip_pat or use_txt)
                prev = t_prev
                # print(op)
        return op[0]
                
    def isMatch(self, text, pattern):
        t_p = self.compress_pattern(pattern)
        return self.is_matching(text, t_p)